File: app/src/main/java/org/blitzortung/android/preferences/SlidePreferences.java
package org.blitzortung.android.preferences;

import android.content.Context;
import android.preference.DialogPreference;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.SeekBar;
import android.widget.TextView;


public class SlidePreferences extends DialogPreference implements SeekBar.OnSeekBarChangeListener {
    private static final String ATTRIBUTE_NAMESPACE = "http://schemas.android.com/apk/res/android";

    private final Context context;
    private TextView valueText;
    private SeekBar slider;

    private final String unitSuffix;

    private final int defaultValue;
    private final int maximumValue;
    private int currentValue;

    public SlidePreferences(Context context, AttributeSet attrs) {
        super(context, attrs);

        this.context = context;

        unitSuffix = " " + attrs.getAttributeValue(ATTRIBUTE_NAMESPACE, "text");
        defaultValue = attrs.getAttributeIntValue(ATTRIBUTE_NAMESPACE, "defaultValue", 30);
        maximumValue = attrs.getAttributeIntValue(ATTRIBUTE_NAMESPACE, "max", 80);
    }

    @Override
    protected View onCreateDialogView() {
        LinearLayout layout = new LinearLayout(context);
        layout.setOrientation(LinearLayout.VERTICAL);
        layout.setPadding(5, 5, 5, 5);

        valueText = new TextView(context);
        valueText.setGravity(Gravity.CENTER_HORIZONTAL);
        valueText.setTextSize(32);
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT);
        layout.addView(valueText, params);

        slider = new SeekBar(context);
        slider.setOnSeekBarChangeListener(this);
        layout.addView(slider, params);

        if (shouldPersist()) {
            currentValue = getPersistedInt(defaultValue);
        }

        slider.setMax(maximumValue);
        slider.setProgress(currentValue);

        return layout;
    }

    @Override
    protected void onBindDialogView(View v) {
        super.onBindDialogView(v);

        slider.setMax(maximumValue);
        slider.setProgress(currentValue);
    }

    @Override
    protected void onSetInitialValue(boolean should_restore, Object defaultValue) {
        super.onSetInitialValue(should_restore, defaultValue);

        if (should_restore) {
            currentValue = shouldPersist() ? getPersistedInt(this.defaultValue) : this.defaultValue;
        } else {
            currentValue = (Integer) defaultValue;
        }
    }

    public void onProgressChanged(SeekBar seekBar, int currentValue, boolean fromTouch) {
        String t = String.valueOf(currentValue);

        valueText.setText(unitSuffix == null ? t : t.concat(unitSuffix));

        if (shouldPersist()) {
            persistInt(currentValue);
        }

        callChangeListener(currentValue);
    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {
    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {
    }
}

File: app/src/main/java/org/blitzortung/android/preferences/SlidePreferences.kt
/*

   Copyright 2015 Andreas Würl

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

*/

package org.blitzortung.android.preferences

import android.content.Context
import android.preference.DialogPreference
import android.util.AttributeSet
import android.view.Gravity
import android.view.View
import android.widget.LinearLayout
import android.widget.SeekBar
import android.widget.TextView


class SlidePreferences(context: Context, attrs: AttributeSet) : DialogPreference(context, attrs),
    SeekBar.OnSeekBarChangeListener {
    private val data: SliderData
    private lateinit var valueText: TextView
    private lateinit var slider: SeekBar

    init {
        data = SliderData(
            suffix = attrs.getAttributeValue(ATTRIBUTE_NAMESPACE, "text"),
            default = attrs.getAttributeIntValue(ATTRIBUTE_NAMESPACE, "defaultValue", 30),
            minimum = attrs.getAttributeIntValue(ATTRIBUTE_NAMESPACE, "min", 0),
            maximum = attrs.getAttributeIntValue(ATTRIBUTE_NAMESPACE, "max", 80),
            step = attrs.getAttributeIntValue(null, "step", 1)
        )
    }

    @Deprecated("Deprecated in Java")
    override fun onCreateDialogView(): View {
        val layout = LinearLayout(context)
        layout.orientation = LinearLayout.VERTICAL
        layout.setPadding(5, 5, 5, 5)

        val valueText = TextView(context)
        valueText.gravity = Gravity.CENTER_HORIZONTAL
        valueText.textSize = 32f
        //Set initial text
        valueText.text = data.text
        val params = LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT,
            LinearLayout.LayoutParams.WRAP_CONTENT
        )
        layout.addView(valueText, params)
        this.valueText = valueText

        slider = SeekBar(context).apply {
            setOnSeekBarChangeListener(this@SlidePreferences)
            layout.addView(this, params)
            max = data.size
            progress = data.offset
        }

        return layout
    }

    @Deprecated("Deprecated in Java")
    override fun onBindDialogView(v: View) {
        super.onBindDialogView(v)
        slider.max = data.size
        slider.progress = data.offset
    }

    @Deprecated("Deprecated in Java")
    override fun onSetInitialValue(should_restore: Boolean, defaultValue: Any?) {
        super.onSetInitialValue(should_restore, defaultValue)

        if (should_restore && shouldPersist()) {
            data.value = getPersistedInt(data.default)
        } else if (defaultValue is Int) {
            data.value = defaultValue
        }
    }

    override fun onProgressChanged(seekBar: SeekBar, seekBarValue: Int, fromTouch: Boolean) {
        if (!fromTouch) {
            return
        }

        data.offset = seekBarValue
        valueText.text = data.text
        callChangeListener(data.value)
    }


    @Deprecated("Deprecated in Java")
    override fun onDialogClosed(positiveResult: Boolean) {
        if (positiveResult) {
            if (shouldPersist())
                persistInt(data.value)
        } else {
            onSetInitialValue(true, data.default)
        }

        super.onDialogClosed(positiveResult)
    }

    override fun onStartTrackingTouch(seekBar: SeekBar) {
    }

    override fun onStopTrackingTouch(seekBar: SeekBar) {
    }

    companion object {
        private const val ATTRIBUTE_NAMESPACE = "http://schemas.android.com/apk/res/android"
    }
}

